ARCs/arc-draft_extended_desc.md

```diff
@@ -0,0 +1,160 @@
+---
+arc: <to be assigned>
+title: Extended App Description
+description: Adds more information to the ARC4 JSON description
+author: Joe Polny (@joe-p)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-11-14
+requires: 4
+---
+
+
+## Abstract
+This ARC takes the existing JSON description of a contract as described in ARC4 and adds more fields for the purpose of client interaction
+
+## Motivation
+The description provided by ARC4 is missing a lot of critical information that clients should know when interaction with an app.
+
+## Specification
+```ts
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [name: string]: string | StructFields;
+}
+
+/** Describes a single key in app storage */
+interface StorageKey {
+  /** The type of the key. Can be ABI type or named struct */
+  keyType: string;
+  /** The type of the value. Can be ABI type or named struct */
+  valueType: string;
+  /** The key itself, as a byte array */
+  key: number[];
+}
+
+interface StorageMap {
+  /** The type of the key. Can be ABI type or named struct */
+  keyType: string;
+  /** The type of the value. Can be ABI type or named struct */
+  valueType: string;
+  /** The prefix of the map, as a string */
+  prefix: string;
+}
+
+interface Method {
+  /** The name of the method */
+  name: string;
+  /** Optional, user-friendly description for the method */
+  desc?: string;
+  /** The arguments of the method, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: string;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+  }>;
+  /** Information about the method's return value */
+  returns: {
+    /** The type of the return value, or "void" to indicate no return value. */
+    type: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: string;
+    /** Optional, user-friendly description for the return value */
+    desc?: string;
+  };
+  /** an action is a combination of call/create and an OnComplete */
+  actions: {
+    /** OnCompeltes this method allows when appID === 0 */
+    create: ('NoOp' | 'OptIn' | 'DeleteApplication')[];
+    /** OnCompeltes this method allows when appID !== 0 */
+    call: ('NoOp' | 'OptIn' | 'CloseOut' | 'ClearState' | 'UpdateApplication' | 'DeleteApplication')[];
+  };
+  readonly: boolean;
+}
+
+interface Contract {
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
```

Comment:
Given we are taking the opportunity to rejig things can we use `description` rather than `desc` or `descr`. `description` is a pretty standard property name e.g. things like json schema, and I think abbreviations and acronyms obscure understandability, particular for non-native english speakers.

Reply:
(Noting this would be a breaking change against ARC4 unfortunately, maybe we can support either so we can move forward, but have backwards compatibility?)

Reply:
Because this would be such a minor change in the grand scheme of things I'm not really sure it would be worth adding extra complexity/breaking backwards compatability

Reply:
Happy for it to stay with arc4 status quo, but if we didn't want to take a stance towards more broadly understandable language it's a now or never moment.